test/functional/sqlsrv/sqlsrv_ae_datetimes_as_strings.phpt

```diff
@@ -0,0 +1,627 @@
+--TEST--
+Test various date and time types with AE and ReturnDatesAsStrings set to true
+--SKIPIF--
+<?php require('skipif_versions_old.inc'); ?>
+--FILE--
+<?php
+require_once('MsCommon.inc');
+
+// Check for expected errors. These are expected in cases where the dates and
+// times do not comply with ODBC standards.
+// 07006 Restricted data type attribute violation (Conversion failed)
+// 22007 Invalid datetime format (ODBC accepts only a few formats)
+// 22008 Datetime field overflow (Outside range)
+// 22018 Invalid character value for cast specification
+function ExecutePreparedStmt($stmt)
+{
+    if ($stmt == false) {
+        fatalError("Failure in sqlsrv_prepare");
+    } else {
+        $r = sqlsrv_execute($stmt);
+        if ($r == false) {
+            $errors = sqlsrv_errors();
+            if ($errors[0]['SQLSTATE'] != '07006' and
+                $errors[0]['SQLSTATE'] != '22007' and
+                $errors[0]['SQLSTATE'] != '22008' and
+                $errors[0]['SQLSTATE'] != '22018') {
+                print_r($errors);
+                fatalError("Unexpected error");
+            }
+        }
+    }
+}
+
+// Compare dates retrieved from the database with the date used for testing.
+// $expectedDateTime is an array of strings like '2002-01-31 23:59:59.049876'
+// $retrievedDateTime is date/time string of format either 'Y-m-d H:i:s.u' or
+// 'Y-m-d H:i:s.u P', which is the format returned when $returnDatesAsStrings 
+// is true, or by the date_format() calls in FetchDatesAndOrTimes when a PHP 
+// DateTime object is retrieved (i.e. when $returnDatesAsStrings is false),
+// unless $datetimetype is date or time. In those cases:
+// If $datetimetype is date and $returnDatesAsStrings is false:
+//   The date is as expected, the time defaults to 00:00:00.0000
+// If $datetimetype is time and $returnStrings is false:
+//   The date defaults to the current date, the time is as expected. 
+// If $datetimetype is date and $returnStrings is true:
+//   $retrievedDateTime is only a date.
+// If $datetimetype is 'time' and $returnStrings is true:
+//   $retrievedDateTime is only a time.
+function CompareDateTime($datetimetype, $returnStrings, &$expectedDateTime, $retrievedDateTime)
```

Comment:
I agree with Jenny. Would be nice if this function is split into 2: one fore CompareDateTimeString, and one for CompareDateTimeObject

Reply:
I'll split it into two, but is it clear that when a DateTime object is returned, the test converts it to a string using date_format() before passing to CompareDateTime?

Reply:
Yes that's clear. Both compare with strings, but one from DatetimeAsString, and one is from a datetime object using date_format()